guides/source/active_record_encryption.md

```diff
@@ -328,281 +408,368 @@ class User
-class Attribute
-  encrypts :title, encryptor: MyAttributeEncryptor.new
+class Article < ApplicationRecord
+  encrypts :summary, key: "some secret key for article summaries"
 end
 ```
 
-#### Encryption Context When Running a Block of Code
+Active Record will use the key passed to `encrypts` to encrypt and decrypt the
+`summary` attribute above.
 
-You can use `ActiveRecord::Encryption.with_encryption_context` to set an encryption context for a given block of code:
+### Rotating Keys
 
-```ruby
-ActiveRecord::Encryption.with_encryption_context(encryptor: ActiveRecord::Encryption::NullEncryptor.new) do
-  # ...
-end
-```
+Active Record Encryption can work with lists of keys to support implementing key
+rotation schemes. The reason to rotate keys may be as part of your
+organization's security policy or if you suspect a key may be compromised.
 
-#### Built-in Encryption Contexts
+In the example below, the *last key* is used for encrypting new content and all
+keys are tried when decrypting content until one works.
 
-##### Disable Encryption
+```yml
+active_record_encryption:
+  primary_key:
+    - a1cc4d7b9f420e40a337b9e68c5ecec6 # Previous keys can still decrypt existing content
+    - bc17e7b413fd4720716a7633027f8cc4 # Active, encrypts new content
+  key_derivation_salt: a3226b97b3b2f8372d1fc6d497a0c0d3
+```
 
-You can run code without encryption:
+This enables key rotation workflow where you keep a short list of keys by adding
+new keys, re-encrypting content, and deleting old keys.
 
-```ruby
-ActiveRecord::Encryption.without_encryption do
-  # ...
-end
-```
+NOTE: Rotating keys is not supported for deterministic encryption.
 
-This means that reading encrypted text will return the ciphertext, and saved content will be stored unencrypted.
+### Storing Key References
 
-##### Protect Encrypted Data
+You can store a reference to the encryption key in the encrypted message itself.
+The advantage of doing this is that decryption can be more performant as the
+system does not have to try a list of keys to find one that works. The tradeoff
+is that the encryption data will be a bit larger.
 
-You can run code without encryption but prevent overwriting encrypted content:
+In order to store a key reference, you need to enable this configuration:
 
 ```ruby
-ActiveRecord::Encryption.protecting_encrypted_data do
-  # ...
-end
+config.active_record.encryption.store_key_references = true
 ```
 
-This can be handy if you want to protect encrypted data while still running arbitrary code against it (e.g. in a Rails console).
+## Configuration
+
+You can configure Active Record Encryption options in your `application.rb`
+(most common scenario) or in a specific environment config file
+`config/environments/<env name>.rb` if you want to set them on a per-environment
+basis. You can see the Active Record Encryption related configuration properties
+in the [Configuration
+Guide](configuring.html#config-active-record-encryption-support-unencrypted-data)
+
```

Comment:
I think we can remove this section now.
It's pretty generic and other guides don't have a configuration section either .
```suggestion
```